source/module_hsolver/genelpa/CMakeLists.txt

```diff
@@ -0,0 +1,20 @@
+#cmake_minimum_required(VERSION 2.8)
+#PROJECT (TEST)
+
+#SET(ELPA_INCLUDE_DIR /home/nic/yshen/program/elpa/2016.05.004/include/elpa-2016.05.004/)
+#SET(ELPA_INCLUDE_DIR /home/nic/yshen/program/elpa/2021.11.002-A100/include/elpa-2021.11.002)
+MESSAGE(STATUS "elpa include dir:" ${ELPA_INCLUDE_DIR} )
+string(REGEX REPLACE "/$" "" ELPA_INC_STR ${ELPA_INCLUDE_DIR})
+string(REPLACE "/" ";" ELPA_INC_STR ${ELPA_INC_STR})
+list(GET ELPA_INC_STR -1 ELPA_VER)
+string(REPLACE "elpa-" "" ELPA_VER ${ELPA_VER})
+message(STATUS "elpa version is: " ${ELPA_VER})
+
+if(${ELPA_VER} VERSION_GREATER "2016.05.004")
+    execute_process(COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/translate_elpa_generic.sh ${ELPA_INCLUDE_DIR})
+    add_library(genelpa OBJECT elpa_new.cpp elpa_new_real.cpp elpa_new_complex.cpp utils.cpp)
+else()
+    execute_process(COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/translate_elpa_generated.sh ${ELPA_INCLUDE_DIR})
```

Comment:
Is it possible to use macros to determine and select the correct interface for elpa, rather than use plain-text replacing?

Reply:
the header files of ELPA follow C standards but some features are not well compatible with C++.

For example, the `_Generic` keyword should be reimpliment in C++ by overload functions.

Another one is the `complex` keyword, defined in C's standard header `complex.h`, is not compatible with C++. One can use `#define complex _Complex` to resolve this problem, but this does NOT work every time for every versions of compiler.

In short, The ELPA does not provide the "correct" interfaces working with standard C++, therefore it's better to create the new header files for C++ rather than try to make the C header files work with C++.

Reply:
@pplab Sorry, let me explain my concern again:
What I'm suggesting is avoid duplicated `elpa_generic_template_*.hpp`. Personally I think in the form of 
```c++
#include <elpa_version.h>
static inline void elpa_eigenvalues(elpa_t handle, float complex *a, float *ev, int *error)
{
#if ELPA_API_VERSION >= 20211125
	elpa_eigenvalues_fc(handle, a, ev, error);
#else 
	elpa_eigenvalues_all_host_arrays_d(handle, a, ev, error);
}
```
should be better. What do you think?

Reply:
The complex compatibility between c and c++ is really bad. I tried to figure out a more general solution but failed, unless compiled separately. Let's leave it as is.

Reply:
> 
This approach would certainly work, and the code would be more streamlined.

There are two reasons for not using this approach. First, I don't know if ELPA team will make major (e.g. structural) changes to this header file in the future to make this conditional compilation less easy to modify; second, inserting a lot of conditional compilation instructions into the code would make me feel like I am always interrupted when reading the code, so I personally don't like this approach.